src/sentry/rules/actions/utils.py

```diff
@@ -31,41 +31,38 @@ def check_value_changed(
 
 
 def generate_diff_labels(
-    present_state: dict[str, Any],
-    prior_state: dict[str, Any],
+    present_state: list[dict[str, Any]],
+    prior_state: list[dict[str, Any]],
     rule: Rule,
-    changed_data: DefaultDict[str, list[str]],
     key: str,
 ) -> DefaultDict[str, list[str]]:
     added_statement = "Added {} '{}'"
     removed_statement = "Removed {} '{}'"
     added_items = []
-
+    changed_data: DefaultDict[str, list[str]] = defaultdict(list)
     # Added items that are not seen in the prior rule data
-    for data in prior_state.get(key, []):
-        if data not in present_state.get(key, []):
+    for data in prior_state:
+        if data not in present_state:
             label = generate_rule_label(rule.project, rule, data)
-            changed_data[data["id"]].append(added_statement.format(key[:-1], label))
+            changed_data[data["id"]].append(added_statement.format(key, label))
             added_items.append(data["id"])
```

Comment:
are `id`s unique for each item? Or could there possibly be 2 `data`'s with the same id?

Rather than looping - we could just construct sets of the ids
Then we can take difference to determine the new ids?

Then if we've constructed a `prior_items_by_id`, we can simply grab the items by id

Reply:
It's possible there could be multiple of the same id, though it's unlikely. We don't stop you from doing something like this: 
<img width="931" alt="Screenshot 2024-03-28 at 11 17 33 AM" src="https://github.com/getsentry/sentry/assets/29959063/ec3d4d05-df1a-49f1-ba4b-f5347b861d66">